lib/r10k/tarball.rb

```diff
@@ -0,0 +1,257 @@
+require 'digest'
+require 'fileutils'
+require 'find'
+require 'minitar'
+require 'net/http'
+require 'tempfile'
+require 'uri'
+require 'zlib'
+require 'r10k/settings'
+require 'r10k/settings/mixin'
+require 'r10k/util/platform'
+
+module R10K
+  class Tarball
+
+    include R10K::Settings::Mixin
+
+    def_setting_attr :proxy
+    def_setting_attr :cache_root, if R10K::Util::Platform.windows?
+                                    File.join(ENV['LOCALAPPDATA'], 'r10k', 'tarball')
+                                  else
+                                    File.expand_path(ENV['HOME'] ? '~/.r10k/tarball': '/root/.r10k/tarball')
+                                  end
+
+    # @!attribute [r] name
+    #   @return [String] The tarball's name
+    attr_accessor :name
+
+    # @!attribute [r] source
+    #   @return [String] The tarball's source
+    attr_accessor :source
+
+    # @!attribute [r] sha256digest
+    #   @return [String] The tarball's expected sha256 digest
+    attr_accessor :sha256digest
+
+    # @param name [String] The name of the tarball content
+    # @param source [String] The source for the tarball content
+    # @param sha256digest [String] The sha256 digest of the tarball content
+    def initialize(name, source, sha256digest: nil)
+      @name = name
+      @source = source
+      @sha256digest = sha256digest
+    end
+
+    # @return [String] Directory. Where the cache will be created.
+    def cache_root
+      File.join(settings[:cache_root], sanitized_dirname(source))
+    end
+
+    # @return [String] File. The path the tarball will be cached to.
+    def cache_path
+      File.join(cache_root, tarball_basename)
+    end
+
+    # @return [String] The basename of the tarball cache file.
+    def tarball_basename
+      name + '.tar.gz'
+    end
+
+    # Extract the cached tarball to the target directory.
+    #
+    # @param target_dir [String] Where to unpack the tarball
+    def unpack(target_dir)
+      file = File.open(cache_path, 'rb')
+      reader = Zlib::GzipReader.new(file)
+      begin
+        Minitar.unpack(reader, target_dir)
+      ensure
+        reader.close
+      end
+    end
+
+    # @param target_dir [String] The directory to check if is insync with the
+    #        tarball content
+    #
+    # @return [Boolean]
+    def insync?(target_dir, purge: true)
```

Comment:
```suggestion
    def in_sync?(target_dir, purge: true)
```
don't love smushing words together 😅

Reply:
Yeah. I tend to agree, but `insync` as a term seems to be the r10k convention. As such, I think it's maybe best to leave it this way in service to the greater goal of consistency.